File: apps/siakad-tu/src/domain/student/Enrollment.js
// apps/siakad-tu/src/domain/student/Enrollment.js

/**
 * Entity: Enrollment
 *
 * Merepresentasikan riwayat pendaftaran siswa di suatu kelas pada periode tertentu.
 * Enrollment adalah bagian dari Student Aggregate.
 *
 * Identity: enrollmentId
 * Lifecycle: terikat pada Student
 */
export class Enrollment {
  #enrollmentId
  #schoolId // dari OperationalContext
  #periodId // dari OperationalContext
  #classId
  #enrollmentDate
  #status // 'active', 'completed', 'transferred'

  constructor({
    enrollmentId,
    schoolId,
    periodId,
    classId,
    enrollmentDate = new Date(),
    status = 'active',
  }) {
    if (!enrollmentId) throw new Error('Enrollment: enrollmentId wajib.')
    if (!schoolId) throw new Error('Enrollment: schoolId wajib.')
    if (!periodId) throw new Error('Enrollment: periodId wajib.')
    if (!classId) throw new Error('Enrollment: classId wajib.')

    this.#enrollmentId = enrollmentId
    this.#schoolId = schoolId
    this.#periodId = periodId
    this.#classId = classId
    this.#enrollmentDate =
      enrollmentDate instanceof Date ? enrollmentDate : new Date(enrollmentDate)
    this.#status = status
  }

  get enrollmentId() {
    return this.#enrollmentId
  }
  get schoolId() {
    return this.#schoolId
  }
  get periodId() {
    return this.#periodId
  }
  get classId() {
    return this.#classId
  }
  get enrollmentDate() {
    return this.#enrollmentDate
  }
  get status() {
    return this.#status
  }

  isActive() {
    return this.#status === 'active'
  }

  /**
   * Cek apakah enrollment ini untuk periode yang sama dengan yang diberikan
   */
  isForPeriod(periodId) {
    return this.#periodId === periodId
  }

  /**
   * Complete enrollment (saat siswa lulus atau pindah)
   */
  complete() {
    if (this.#status !== 'active') {
      throw new Error(
        `Enrollment ${this.#enrollmentId} tidak bisa di-complete. Status saat ini: ${this.#status}`,
      )
    }
    this.#status = 'completed'
  }

  toJSON() {
    return {
      enrollmentId: this.#enrollmentId,
      schoolId: this.#schoolId,
      periodId: this.#periodId,
      classId: this.#classId,
      enrollmentDate: this.#enrollmentDate.toISOString(),
      status: this.#status,
    }
  }

  static fromJSON(data) {
    return new Enrollment(data)
  }
}

File: services/tu-core/internal/domain/student/models/enrollment.go
package models

import "time"

// Enrollment adalah Entity dalam Student Aggregate.
// SYMMETRIC dengan Enrollment.js di frontend.
type Enrollment struct {
	enrollmentID   string
	schoolID       string
	periodID       string
	classID        string
	enrollmentDate time.Time
	status         string // active, completed, transferred
}

// NewEnrollment membuat Enrollment baru.
func NewEnrollment(enrollmentID, schoolID, periodID, classID string) *Enrollment {
	return &Enrollment{
		enrollmentID:   enrollmentID,
		schoolID:       schoolID,
		periodID:       periodID,
		classID:        classID,
		enrollmentDate: time.Now().UTC(),
		status:         "active",
	}
}

// RehydrateEnrollment merekonstruksi Enrollment dari data persisten.
func RehydrateEnrollment(enrollmentID, schoolID, periodID, classID, status string, enrollmentDate time.Time) *Enrollment {
	return &Enrollment{
		enrollmentID:   enrollmentID,
		schoolID:       schoolID,
		periodID:       periodID,
		classID:        classID,
		enrollmentDate: enrollmentDate,
		status:         status,
	}
}

// Getters

func (e *Enrollment) EnrollmentID() string      { return e.enrollmentID }
func (e *Enrollment) SchoolID() string          { return e.schoolID }
func (e *Enrollment) PeriodID() string          { return e.periodID }
func (e *Enrollment) ClassID() string           { return e.classID }
func (e *Enrollment) EnrollmentDate() time.Time { return e.enrollmentDate }
func (e *Enrollment) Status() string            { return e.status }

// IsActive mengecek apakah enrollment aktif.
func (e *Enrollment) IsActive() bool {
	return e.status == "active"
}

// IsForPeriod mengecek apakah enrollment untuk periode tertentu.
func (e *Enrollment) IsForPeriod(periodID string) bool {
	return e.periodID == periodID
}

// Complete menandai enrollment sebagai selesai.
func (e *Enrollment) Complete() {
	if e.status != "active" {
		return
	}
	e.status = "completed"
}
